src/hotspot/share/gc/z/zRelocate.cpp

```diff
@@ -795,6 +798,14 @@ class ZRelocateWork : public StackObj {
     update_remset_promoted(to_addr);
   }
 
+  void maybe_string_dedup(zaddress to_addr) {
+    const bool is_promotion = _forwarding->to_age() == ZPageAge::old && _forwarding->from_age() != ZPageAge::old;
```

Comment:
It seems like this could be computed once at `ZRelocateWork` construction time?  Unless `_forwarding`
can change.  Also, isn't this the same as `_forwarding->is_promotion()`?

Reply:
`_forwarding` changes whenever we move to the next ZPage to relocate. We could probably add an `_is_promotion` variable/constant to ZForwarding, but we're already reading `_to_age` and `_from_age` in `update_remset_for_fields` so I don't expect a noticeable performance difference if we would do that.

Great point about `_forwarding->is_promotion()`. I'm updating the code to use that.